src/boot/efi/pe.c

```diff
@@ -160,6 +166,49 @@ static void locate_sections(
         }
 }
 
+static UINT32 get_compatibility_entry_address(const struct DosFileHeader *dos, const struct PeFileHeader *pe) {
+        UINTN addr = 0, size = 0;
+        static const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
+
+        /* The kernel may provide alternative PE entry points for different PE architectures. This allows
+         * booting a 64bit kernel on 32bit EFI that is otherwise running on a 64bit CPU. The locations of any
+         * such compat entry points are located in a special PE section. */
+
+        locate_sections((const struct PeSectionHeader *) ((const UINT8 *) dos + section_table_offset(dos, pe)),
+                        pe->FileHeader.NumberOfSections,
+                        sections,
+                        &addr,
+                        NULL,
+                        &size);
+
+        if (size == 0)
+                return 0;
+
+        typedef struct {
+                UINT8 type;
+                UINT8 size;
+                UINT16 machine_type;
+                UINT32 entry_point;
+        } _packed_ LinuxPeCompat1;
+
+        while (size >= sizeof(LinuxPeCompat1) && addr % __alignof__(LinuxPeCompat1 *) == 0) {
```

Comment:
actually, must be `__alignof__(LinuxPeCompat1)`, i.e. of the struct, not the pointer.

might make sense to wrap is in a helper macro, something like this:

```c
#define POINTER_IS_ALIGNED(p, t) ({ \
        const void *_q = (p); \
        UINTN a = (UINTN) _q; \
        (a % __alignof__(t)) == 0; \
})
```

With that you could just do:

```c
while (size >= sizeof(LinuxPeCompat1) && POINTER_IS_ALIGNED((UINT8*) dos + addr, LinuxPeCompat1) {
…
``

Reply:
Huh. Interestingly, alignof the _packed_ struct is 1, while 4 without. Does checking really make sense then??

Reply:
Yeah, the _packed_ removes the alignment requirements, that's why I asked about that earlier.

But I'd still do the check, since in general we should check alignment before assuming too much from arbitrary pointers.

Reply:
alternatively, add an `assert_cc(__alignof__(LinuxPeCompat1) == 1)` before casting into a pointer of that type. That documents clearly that the assumption here is that any pointer is safe for this structure.